### ISSUE TITLE
Avoid_toList_on_spreads lint request

### ISSUE THREAD

**Issue by mit-mit** [dart-lang/linter#2965]
Lots of Dart UI code operates on lists (e.g. a list of Widgets in Flutter). As some core library functions intern iterables, some developers have a habit of calling `toList()` on those without reflecting on it. One example I often see is doing this in conjunction with the spread operator, where calling toList is not needed, as the spread can be called on the iterable directly.

**Examples**

BAD:
```dart
  children: <Widget>[
    ...['foo', 'bar', 'baz'].map((String s) => Text(s)).toList(),
  ]
```

GOOD:
```dart
  children: <Widget>[
    ...['foo', 'bar', 'baz'].map((String s) => Text(s)),
  ]
```

**Comment by lrhn** [comment]
In most places where you immediately iterate the result, calling `.toList()` is unnecessary (`.toSet()` is not, it potentially deduplicates). It does indeed do an extra allocation and creates a list which is then iterated once and discarded.

I think "immediately iterates" is currently just spreads and `for`/`in`, and it's a little iffy for `for`/`in` because you there are situations where it does make sense to make a copy of the iterable: When the later code might (concurrently) change the thing that would otherwise be iterated and break the iteration.

Say, something like:
```dart
    static void transitivelyRemove<T>(Map<T, List<T>> map, T key) {
      var subKeys = map.remove(key);
      if (subKeys != null) {
        for (var subKey in subKeys) {
          transitivelyRemove(map, subKey);
        }
      }
    }
    ...
      for (var key in map.keys.toList()) {
        if (map.containsKey(key) && shouldTransitivelyRemove(key)) {
          transitivelyRemove(map, key);
        }
      }
```
So, hinting that a `.toList` on a `for`/`in` iteration is unnecessary isn't necessarily true. I don't know if there are easy heuristics to rule out the `.toList()` being needed, but ... probably not.

I think it'll be _safe enough_ to do it on spreads. Maybe also in a `for`/`in` inside a collection literal, since the element expressions of collection literals are usually not having any side effects. And if they do, you should probably reconsider what you're doing. (Then do it anyway because it's cool, but you have to write the `//ignore`!)

It could probably be extended to immediately calling `.forEach` on the `iterable.toList()`, or any other iterable method, but there I'd actually expect the `toList()` to be deliberate. People are used to doing `toList()` at the end of an iterable method chain, I don't envision them putting it in the middle by accident very often.

**Comment by pq** [comment]
@jakemac53 @natebosch @munificent : this feels like a reasonable core lint but I think we need to get the scope just right.  Would love to hear your thoughts.

**Comment by munificent** [comment]
I agree with Lasse. I think linting on `toList()` inside a spread is safe and a good idea, but linting on `toList()` inside for-in loop will likely have false positives. (I know I have code that deliberately does `toList()` inside a for-in loop.)

**Comment by mit-mit** [comment]
@pq any news about this one? If not, i'd like to nominate this for q1?

**Comment by pq** [comment]
>  i'd like to nominate this for q1?

SGTM

**Comment by pq** [comment]
My thinking is to push ahead on a lint scoped to spreads.  Spitballing names, how about `unnecessary_to_list_in_spreads`?

**Comment by mit-mit** [comment]
@pq any news about this one?

**Comment by pq** [comment]
Hey @mit-mit.  This is still in the queue.

Maybe we can get some more feedback from core lints folks?  If there's a consensus this should be suggested style, it'd help bump the priority.

/cc @natebosch @munificent @jakemac53 @lrhn @goderbauer

**Comment by goderbauer** [comment]
Sounds like a useful lint for "recommended" to me.

**Comment by jakemac53** [comment]
+1 for linting on `toList()` in a spread
